repo2docker/contentproviders/dataverse.py

```diff
@@ -0,0 +1,129 @@
+import os
+import json
+import shutil
+
+from urllib.request import Request
+from urllib.parse import urlparse, urlunparse, parse_qs
+
+from .doi import DoiProvider
+from ..utils import copytree, deep_get
+
+
+class Dataverse(DoiProvider):
+    """Provide contents of a Dataverse dataset."""
+
+    def __init__(self):
+        data_file = os.path.join(os.path.dirname(__file__), "dataverse.json")
+        with open(data_file, "r") as fp:
+            self.hosts = json.load(fp)["installations"]
+        super().__init__()
+
+    def detect(self, doi, ref=None, extra_args=None):
+        """Trigger this provider for things that resolve to a Dataverse dataset.
+
+        Handles:
+        - DOI pointing to {siteURL}/dataset.xhtml?persistentId={persistentId}
+        - DOI pointing to {siteURL}/file.xhtml?persistentId={persistentId}&...
+        - URL {siteURL}/api/access/datafile/{fileId}
+
+        Examples:
+        - https://dataverse.harvard.edu/api/access/datafile/3323458
+        - doi:10.7910/DVN/6ZXAGT
+        - doi:10.7910/DVN/6ZXAGT/3YRRYJ
+
+        """
+        url = self.doi2url(doi)
+        # Parse the url, to get the base for later API calls
+        parsed_url = urlparse(url)
+
+        # Check if the url matches any known Dataverse installation, bail if not.
+        host = next(
+            (
+                host
+                for host in self.hosts
+                if urlparse(host["url"]).netloc == parsed_url.netloc
+            ),
+            None,
+        )
+        if host is None:
+            return
```

Comment:
To check I parsed the above lines properly (the use of `next()` to handle the "didn't work" case made me read it a few times), is the following code equivalent (I think yes):
```python
for host_ in self.hosts:
    if urlparse(host_["url"]).netloc == parsed_url.netloc:
        host = host_
else:
    return
```